packages/flutter_tools/lib/src/base/file_system.dart

```diff
@@ -17,6 +17,40 @@ import 'signals.dart';
 // the tool is killed by a signal.
 export 'package:file/file.dart';
 
+extension FileUtils on File {
+  /// Delete the file and return true if it exists, take no action and return
+  /// false if it does not.
+  ///
+  /// This method should be prefered to checking if it exists and
+  /// then deleting, because it handles the edge case where the file is
+  /// deleted by a different program between the two calls.
+  bool deleteIfExists({bool recursive = false}) {
+    if (!existsSync()) {
+      return false;
+    }
+    try {
+      deleteSync(recursive: recursive);
+    } on FileSystemException catch (err) {
+      // Certain error codes indicate the file could not be found. It could have
+      // been deleted by a different program while the tool was running.
+      // if it still exists, the file likely exists on a read-only volume.
+      // This error code is the same across linux, windows, and macOS.
+      const int kSystemCannotFindFile = 2;
+      if (err?.osError?.errorCode != kSystemCannotFindFile) {
+        rethrow;
+      }
+      if (existsSync()) {
+        throwToolExit(
```

Comment:
This would subvert a `noExitOnFailure`.

Reply:
Thats true - this should probably become a static method on error handling file system - then that can be taken into account.

Reply:
Updated to support noExitOnFailure